enhancements/scheduling/load-aware-vm-rebalancing-through-descheduler.md

```diff
@@ -0,0 +1,472 @@
+* **Hard-tainting**: hard taints (`effect: NoSchedule`/`effect: NoExecute`)  can restrict scheduling decisions but may also reduce overall cluster flexibility. The node controller already applies certain condition-based hard taints for critical node states such as `MemoryPressure`, `DiskPressure`, and `PIDPressure`. However, these hard taints are only set when a node reaches a serious or critical condition. The goal of this proposal is to adopt a similar approach using a soft taint when a node is overutilized but not yet in a critical state. This ensures that the scheduler is aware of resource pressure without making the node entirely unschedulable.
+
+## Proposal
+
+The enhancements will include the following functionality changes:
+1. Updating the descheduler to be load-aware to evict pods based on actual resource utilization.
+2. Soft-tainting over-utilized nodes so the requested resources are less effective during scheduling.
+3. Extending the descheduler operator customization to enable the load-aware eviction with tuning options.
+
+### Update LowNodeUtilization plugin to consume Prometheus metrics
+
+Currently, the descheduler's `LowNodeUtilization` plugin classifies node utilization based on Kubernetes metrics. However, these metrics primarily capture CPU and memory usage at the node and pod levels, which are not well-suited for VMs. By consuming Prometheus metrics, the plugin can leverage a broader range of data that better aligns with VM-specific needs.
+
+The existing plugin code is already structured to support additional resource clients, making integration straightforward. To simplify implementation, the plugin will accept a single metric entry per node, with values in the <0;1> interval—where higher value indicates higher node utilization. Since resource thresholds are expressed as percentages by default, this classification remains intuitive and easy to interpret.
+
+In the current iteration of the design, PSI pressure metrics were selected. Due to their nature, it is not possible to predict how evicting a single VM will impact a node's reported PSI pressure. Therefore, the plugin will be designed to evict only one VM per over-utilized node per cycle by default with option to evict more. Since it is safe to assume that each evicted VM will undergo live migration and the number of evictions per node is limited, running the descheduling cycle more frequently (e.g., every 3 minutes) will approximate the effect of evicting multiple VMs within a single cycle.
```

Comment:
> how evicting a single VM will impact a node's reported PSI

Question:
- is a "single VM" represented by a Pod object?
- what does the Node object represent here, the VM or something else?

Can you please clarify this topology, it would be helpful for someone not familiar with this domain